Code Analysis - Improvements to C++ Script & Support for Assembly Keywords

Dalevyy opened the pull request:
# Summary
This PR addresses and closes #78 and #79.

# Description
This PR updates the script presented in #76 based on the suggestions that @ortegj8 and @eliJerald made to me. 

The most notable change is how the output JSON file is generated which can be seen below. Instead of individually denoting each keyword found, I have created a new key called "wordsFound" and inside this key, we will be storing a keyword IF it has been seen at least ONCE. From there, the keyword itself will act as a key to give information such as the amount of times it has been seen and the starting positions it was found in. I have also created a new key called "totalFound" that will keep track of all banned keywords that were encountered in case we need the front end to display that information. Overall, the output JSON file is much cleaner and easier to work with, so I hope the front end will be happy with this data.

The script still does not support interfacing with the database since, as of the moment, I am still unsure about when and how this data will be stored by the front end, and no one has taken on #66. 

Now this script itself is not the most useful since CS135 professors have already taken this into account (albeit in a less sophisticated manner); however, I have created a separate script to handle CS218 assignments since TAs are manually checking themselves if students are using instructions they shouldn't be. As a result, I have moved pieces of my original script into a separate `KeywordAnalyzer` script that will handle all the file I/O, and it will be responsible for calling the correct analyzer.

For CS218, I couldn't find any packages that can handle the syntax analysis of x86 and MIPS code, so I have opted to create my own syntax analysis. This PR currently includes classes for Tokens and an assembly Lexer that will generate x86 and MIPS tokens that will then be used for the analysis of code. This could further be expanded to create an AST (or realistically a linked list structure) for even further analysis; however, I kept it simple and I'm just using tokens to help me analyze the code for blacklisted keywords. Currently, the lexer is generating the correct tokens, and I am getting the correct output in the JSON, so I would say it's successful. 😄  

# Documentation
- New Files Added
    - `KeywordAnalyzer.py`: This is a new central script that the front end will interface with to perform any code analysis. Some of the helper methods related to file I/O have been moved to this script from #76.
    - `cpp`: Package that will store C++ code analysis files
        - `cppCppAnalyzer.py`: This is the script from #76, but it has now been simplified to not include any file I/O handling. Additionally, I figured out how to generate JSON data, so this file is cleaner to read.
    - 'asm': A new package to store all assembly-related code analysis.
        - `AsmAnalyzer.py`: This is the main script that will handle the tokenization of assembly code and generate the correct data needing to be outputted into the JSON file.
        - `asm_lexer`: Package to store Lexer
            - ` Lexer.py`: This script contains the main code for the assembly lexer. A good chunk of the script is devoted to the tokenization of assembly instructions, but it's just a basic lexer.
        - `asm_token`: Package to store code related to generating tokens for the assembly lexer.
            -  `Position.py`: A class that keeps track of the position of a token.
            -  `Token.py`: A class that generates tokens. A token will have a type, a lexeme, and a starting/ending position.
            - `TokenType.py`: An enumeration that keeps track of all types of tokens that can be generated by the assembly lexer.

- Input
The input file is still the same, but now the title should be `CS<className>_AS<assignmentNum>_Words.txt` since we need to make sure we check the correct class data.
 
- Output
This is an updated image for how the output JSON file will be formatted: 
![image](https://github.com/user-attachments/assets/098c2d0f-5eef-4403-af40-802e0d3d1878)

Additionally, the output file is now titled as `<className>_<assignmentNum>_Found_Words.json`.

# Unit Tests
**No tests.**

I have only done manual input testing, but I have not written tests for this PR either. I am thinking about adding unit tests in my next PR depending on what feature I will work on next.

# PR Reviews
I would like at least @ortegj8 to take a look at the changes I made, and I'm fine with anyone else reviewing this PR. 🙂

Dalevyy committed:
Added support for checking for header quotes in include statements..

Dalevyy committed:
Created initial outline for assembly keyword analysis script.

Dalevyy committed:
Created input file validation which is similar to previous script.

Dalevyy committed:
Created Position class to keep track of asm file positions.

Dalevyy committed:
Created initial TokenType Enum to keep track of the types of tokens we will have.

Dalevyy committed:
Created Token class that will assist in the asm lexer.

Dalevyy committed:
Saving init files into commit history.

Dalevyy committed:
Create initial outline and helper methods for lexer.

Dalevyy committed:
Added helper methods to deal with whitespace and new line characters.

Dalevyy committed:
Removed extra constructor.

Dalevyy committed:
Fixed constructor and toString issues.

Dalevyy committed:
Fixed support for EOF and added testing method for assistance.

Dalevyy committed:
Added tokenization support for initial operators and delimiters.

Dalevyy committed:
Added the ability to tokenize numerical values.

Dalevyy committed:
Added support for tokenizing a string literal.

Dalevyy committed:
Created method to tokenize all names and beginning to handle x86 instructions

Dalevyy committed:
Added token types for all x86 instructions and general purpose registers.

Dalevyy committed:
Added generation of tokens for all 218-related x86 instructions and registers.

Dalevyy committed:
Added tokenization for x86 data types.

Dalevyy committed:
Added types for x86 floating point registers and data types.

Dalevyy committed:
Added support to ignore comments.

Dalevyy committed:
Added more support to handle '_' and '.' in names.

Dalevyy committed:
Added extra check for period in names.

Dalevyy committed:
Added support for MIPS instructions discussed in 218.

Dalevyy committed:
Added support for MIPS registers and data types.

Dalevyy committed:
Added header documentation to explain TokenType.

Dalevyy committed:
Added support to tokenize MIPS instructions and registers.

Dalevyy committed:
Added support to ignore MIPS comments.

Dalevyy committed:
Added token for "=" since it is supported in MIPS.

Dalevyy committed:
Created new central script that will interface with C++ and assembly keyword analysis scripts.

Dalevyy committed:
Moved input file validation and opening into central script.

Dalevyy committed:
Refactored input validation method to have more error checks.

Dalevyy committed:
Moved creation of output JSON file into central script.

Dalevyy committed:
Created separate file to handle assembly keyword analysis.

Dalevyy committed:
Removed test code and moving it into AsmAnalyzer instead.

Dalevyy committed:
Removed initial assembly keyword analyzer file.

Dalevyy committed:
Created helper method that will perform the assembly tokenization.

Dalevyy committed:
Passing in assignment number in order to open the correct file

Dalevyy committed:
Added tracking of assembly word instances and line positions.

Dalevyy committed:
Fixed formatting issues with generating JSON file and can now be generated for asm code.

Dalevyy committed:
Removed manual check for file count since this isn't needed anymore.

Dalevyy committed:
Added return statement, so central script can receive data for storing into output file.

Dalevyy committed:
Added separate JSON fields to count all instances of banned keywords and organized the JSON file better.

Dalevyy committed:
Added ability to count total number of keywords after analysis is complete.

Dalevyy committed:
Renamed packaged to be 'asm'.

Dalevyy committed:
Renamed packaged to be 'cpp' and class is now renamed to be 'CppAnalyzer'.

Dalevyy committed:
Removed `runAnalysis` method since it is now handled in the central script.

Dalevyy committed:
Removed saving the JSON output file since it wasn't being used.

Dalevyy committed:
Updating fields to match data passed from central script.

Dalevyy committed:
Refactored header check and simplified saving header info into JSON file.

Dalevyy committed:
Refactored 'checkAST' method to handle new output data formatting for JSON files.

Dalevyy committed:
Reupdated file to check for `main.cpp` instead of asm files.

Dalevyy committed:
Fixed parsing student name from student submission directory.

Dalevyy committed:
Fixed spacing issue with student names.

Dalevyy committed:
Updated checkAST to account for user-defined functions to prevent false positives.

Dalevyy committed:
Adding `code_analysis` packaged to Django settings.

Dalevyy committed:
Refactored code, so it can be reused for both 135 and 218 code analysis.

eliJerald reviewed (approved):
Good work with this Daniel! I like how you added to the readability of your code by adding nodes of tokens, and added separate files/directories, rather than putting everything in just one file. This is really good work!

Thank you for adding the logic for quotation marks for the #include statements. Though it is small, it was important so good job with that! 

I dont see any problems with the PR other than the lint problems but after that the PR should be good to approve and merge.

EvanShumaker commented:
This PR looks like it has a lot of meaningful refactoring and additions. 

I think the keywordAnalyzer.py and 2 packages separated based on language types is a great way to organize it. Also, keeping I/O separate from the analysis logic should make the code easier to test and reuse, also making it all a lot easier to maintain. 

Great work Daniel!

Dalevyy committed:
Added file header.

Dalevyy committed:
Added documentation to AsmAnalyzer class methods.

Dalevyy committed:
Added documentation to Position.py to solve linter errors.

Dalevyy committed:
Added header file for Token class.

Dalevyy committed:
Updated spacing to help make file slightly smaller.

Dalevyy committed:
Added documentation to the class in order to fix linter errors.

Dalevyy committed:
Removed additional new lines to fix the linter errors for this file.

Dalevyy committed:
Added header information for this file.

Dalevyy committed:
Fixed linter issues for this file.

Dalevyy committed:
Trying to fix spacing since linter is still giving errors.

Dalevyy committed:
Added extra new line at start of class definition to finally resolve linter.

Dalevyy committed:
Edited header comment to reflect new changes to script.

Dalevyy committed:
Fixed linter errors for this file.

Dalevyy committed:
Solving last few linter errors, so CppAnalyzer should be free of formatting errors.

Dalevyy committed:
Made line in the "imperative" mood since the linter brought up a new error.

Dalevyy committed:
Spacing issues now resolved as stated by the liner.

Dalevyy committed:
Added documentation to all helper methods in the Lexer class.

Dalevyy committed:
Updated script to try to fix linter errors.

Dalevyy committed:
Fixed final linter errors for AsmAnalyzer.py

Dalevyy committed:
Fixed final linter errors for `Lexer.py`.

Dalevyy committed:
Fixed what should be the last remaining linter errors before PR can be merged.

Dalevyy commented:
I have fixed the majority of linter issues, and this PR is ready to be merged. I have asked @EvanShumaker to review the documentation changes and approve of the merge. Unfortunately, the CI is still failing, but it's failing for the `push` instead of the `pull_request` as seen below. I do not know why this is, but I am honestly annoyed trying to resolve these errors. I am going to ignore it since it contains the warnings we just added to be ignored in #83. If anyone has ideas for why this might be a problem, please go ahead and chime in. 👍 

![image](https://github.com/user-attachments/assets/135b648a-148f-4eaa-a54c-55f4a971c4c9)

EvanShumaker reviewed (approved):
Linter stuff was cleaned up, and my previous comment still applies. These additions all look good to me!

Dalevyy commented:
# MVP Update

This PR is not integrated in our MVP. This is for the same reason as described in #76 where the front end page for this script was not created, so a professor is not able to use this feature with our MVP. Additionally, this script contains code analysis for x86 programs and while our MVP could support x86 programs, our sample dataset currently does not contain x86 programs. We would have to add additional x86 programs for demonstration purposes, and we have decided to keep our PR limited to work with coursework from CS 135.

As a result, this PR remains in our back end file structure waiting to be used.
